### PR TITLE
feat(dependencies): Create script to bundle Java dependencies

### PR DESCRIPTION
**Pull request by bertschneider**
### Motivation/Context

If a customer is using a private registry/repository and (CS) needs to download all dependencies to upload them into this registry, we should help them by providing an easy way to download all (transitive) dependencies.

This came up in a meeting with Jacek Trossen.

### Description

Adds a small Linux/macOS script to download and package all CI Fuzz Java dependencies.

A resulting dependency archive looks something like:

```
» tar tf ci-dependencies.tar
./
./com/
./org/
./org/junit/
./org/junit/jupiter/
./org/junit/jupiter/junit-jupiter-params/
./org/junit/jupiter/junit-jupiter-params/5.9.0/
./org/junit/jupiter/junit-jupiter-params/5.9.0/junit-jupiter-params-5.9.0.pom
./org/junit/jupiter/junit-jupiter-params/5.9.0/junit-jupiter-params-5.9.0.jar
./org/junit/jupiter/junit-jupiter-api/
./org/junit/jupiter/junit-jupiter-api/5.9.0/
./org/junit/jupiter/junit-jupiter-api/5.9.0/junit-jupiter-api-5.9.0.pom
./org/junit/jupiter/junit-jupiter-api/5.9.0/junit-jupiter-api-5.9.0.jar
./org/junit/platform/
./org/junit/platform/junit-platform-launcher/
./org/junit/platform/junit-platform-launcher/1.9.0/
./org/junit/platform/junit-platform-launcher/1.9.0/junit-platform-launcher-1.9.0.pom
./org/junit/platform/junit-platform-launcher/1.9.0/junit-platform-launcher-1.9.0.jar
./org/junit/platform/junit-platform-commons/
./org/junit/platform/junit-platform-commons/1.9.0/
./org/junit/platform/junit-platform-commons/1.9.0/junit-platform-commons-1.9.0.pom
./org/junit/platform/junit-platform-commons/1.9.0/junit-platform-commons-1.9.0.jar
./org/junit/platform/junit-platform-engine/
./org/junit/platform/junit-platform-engine/1.9.0/
./org/junit/platform/junit-platform-engine/1.9.0/junit-platform-engine-1.9.0.pom
./org/junit/platform/junit-platform-engine/1.9.0/junit-platform-engine-1.9.0.jar
./org/junit/junit-bom/
./org/junit/junit-bom/5.9.0/
./org/junit/junit-bom/5.9.0/junit-bom-5.9.0.pom
./org/apiguardian/
./org/apiguardian/apiguardian-api/
./org/apiguardian/apiguardian-api/1.1.2/
./org/apiguardian/apiguardian-api/1.1.2/apiguardian-api-1.1.2.pom
./org/apiguardian/apiguardian-api/1.1.2/apiguardian-api-1.1.2.jar
./org/opentest4j/
./org/opentest4j/opentest4j/
./org/opentest4j/opentest4j/1.2.0/
./org/opentest4j/opentest4j/1.2.0/opentest4j-1.2.0.pom
./org/opentest4j/opentest4j/1.2.0/opentest4j-1.2.0.jar
./
./com/
./com/code-intelligence/
./com/code-intelligence/jazzer-junit/
./com/code-intelligence/jazzer-junit/0.23.0/
./com/code-intelligence/jazzer-junit/0.23.0/jazzer-junit-0.23.0.pom
./com/code-intelligence/jazzer-junit/0.23.0/jazzer-junit-0.23.0.jar
./com/code-intelligence/cifuzz-maven-extension/
./com/code-intelligence/cifuzz-maven-extension/1.2.0/
./com/code-intelligence/cifuzz-maven-extension/1.2.0/cifuzz-maven-extension-1.2.0.pom
./com/code-intelligence/cifuzz-maven-extension/1.2.0/cifuzz-maven-extension-1.2.0.jar
./com/code-intelligence/cifuzz/
./com/code-intelligence/cifuzz/com.code-intelligence.cifuzz.gradle.plugin/
./com/code-intelligence/cifuzz/com.code-intelligence.cifuzz.gradle.plugin/1.11.0/
./com/code-intelligence/cifuzz/com.code-intelligence.cifuzz.gradle.plugin/1.11.0/com.code-intelligence.cifuzz.gradle.plugin-1.11.0.pom
./com/code-intelligence/cifuzz-gradle-plugin/
./com/code-intelligence/cifuzz-gradle-plugin/1.11.0/
./com/code-intelligence/cifuzz-gradle-plugin/1.11.0/cifuzz-gradle-plugin-1.11.0.pom
./com/code-intelligence/cifuzz-gradle-plugin/1.11.0/cifuzz-gradle-plugin-1.11.0.jar
./com/code-intelligence/jazzer-api/
./com/code-intelligence/jazzer-api/0.23.0/
./com/code-intelligence/jazzer-api/0.23.0/jazzer-api-0.23.0.pom
./com/code-intelligence/jazzer-api/0.23.0/jazzer-api-0.23.0.jar
./com/code-intelligence/jazzer/
./com/code-intelligence/jazzer/0.23.0/
./com/code-intelligence/jazzer/0.23.0/jazzer-0.23.0.pom
./com/code-intelligence/jazzer/0.23.0/jazzer-0.23.0.jar
```

### How to use/reproduce

```sh
./tools/dependency-bundler/bundle-dependencies.sh <repository_user> <repository_password> [--clean]
```

### CODE REVIEW

**Comment by bertschneider** [comment]
@JTrossen Is this script sufficient for your needs?

**Review by fmeum — COMMENTED** [review]
The implementation looks good to me, I just wonder whether this is something we should officially maintain in this repo. Wouldn't most private repositories offer some way to request the one-time download (and caching) of a particular dependency? Having a third-party script like this one associate Maven coordinates with essentially arbitrary files doesn't sound like a process we should encourage.

File: tools/dependency-bundler/bundle-dependencies.sh [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+    curl -L 'https://github.com/coursier/coursier/releases/download/v2.1.8/coursier.jar' -o $COURSIER_JAR
+fi
+
+# Add private repository and credentials.
+export COURSIER_REPOSITORIES="central|https://gitlab.code-intelligence.com/api/v4/projects/89/packages/maven"
+export COURSIER_CREDENTIALS="gitlab.code-intelligence.com $1:$2"
+export COURSIER_CACHE="$COURSIER_CACHE"
+
+echo "INFO: Resolving dependencies"
+java -jar $COURSIER_JAR fetch "com.code-intelligence:jazzer-junit:0.23.0"
```
[the remark this answers is not in the document]
Comment (bertschneider) [reply tools/dependency-bundler/bundle-dependencies.sh]: Yes, it does, wasn't aware of it. Changed.

File: tools/dependency-bundler/bundle-dependencies.sh [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+		rm -rf /tmp/ci-coursier
+fi
+
+COURSIER_JAR=/tmp/coursier.jar
+COURSIER_CACHE=/tmp/ci-coursier
+DEPENDENCIES_ARCHIVE=ci-dependencies.tar
+
+if [[ ! -f "$COURSIER_JAR" ]]; then
+    echo "INFO: Could not find \"$COURSIER_JAR\". Downloading..."
+    curl -L 'https://github.com/coursier/coursier/releases/download/v2.1.8/coursier.jar' -o $COURSIER_JAR
```
[the remark this answers is not in the document]
Comment (bertschneider) [reply tools/dependency-bundler/bundle-dependencies.sh:24]: Done.

File: tools/dependency-bundler/bundle-dependencies.sh [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+export COURSIER_CREDENTIALS="gitlab.code-intelligence.com $1:$2"
+export COURSIER_CACHE="$COURSIER_CACHE"
+
+echo "INFO: Resolving dependencies"
+java -jar $COURSIER_JAR fetch "com.code-intelligence:jazzer-junit:0.23.0"
+java -jar $COURSIER_JAR fetch "com.code-intelligence:cifuzz-maven-extension:1.2.0"
+java -jar $COURSIER_JAR fetch "com.code-intelligence.cifuzz:com.code-intelligence.cifuzz.gradle.plugin:1.11.0"
+
+echo "INFO: Packaging dependencies"
+# Remove empty code-intelligence folders in repo1.maven.org cache,
```
[the remark this answers is not in the document]
Comment (bertschneider) [reply tools/dependency-bundler/bundle-dependencies.sh]: Coursier is configured to look up dependencies in `central` and the CI repository. It always creates the respective directory structures regardless if a given dependency can be found in a repository or not. This shouldn't cause any problems down the line.

**Comment by bertschneider** [comment]
Yes, indeed! Having to download all dependencies manually and uploading them, through whatever strange ways, into a private repository sounds like something to avoid at all costs. I also only know private repositories as transparent proxies for other public or private ones, in which case this wouldn't be necessary. 

That being said, having a way to easily download all dependencies is an explicit request from @JTrossen, as apparently some customers won't allow more convenient approaches.

**Review by fmeum — APPROVED** [review]
I do know of the need to do this. Feel free to merge, but an alternative could be to have CS own this as part of their toolset.

**Pull request merged**